px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp

```diff
@@ -0,0 +1,74 @@
+/****************************************************************************
+ * Copyright (c) 2023 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_ros2/common/context.hpp>
+
+namespace px4_ros2
+{
+/** \ingroup odometry
+ *  @{
+ */
+
+/**
+ * Provides a subscription to arbitrary ROS topics.
+ */
+template<typename RosMessageType>
+class Subscription
+{
+public:
+  Subscription(Context & context, const std::string & topic)
+  : _node(context.node()), _topic(context.topicNamespacePrefix() + topic)
+  {
+    _subscription = _node.create_subscription<RosMessageType>(
+      _topic, rclcpp::QoS(1).best_effort(),
+      [this](const typename RosMessageType::UniquePtr msg) {
+        _last = *msg;
+        _last_message_time = _node.get_clock()->now();
+        for (const auto & callback : _callbacks) {
+          callback(_last);
+        }
+      });
+  }
+
+  /**
+   * @brief Register a callback to be executed for each message.
+   *
+   * @param callback the callback function
+   */
+  void subscribe(std::function<void(RosMessageType)> callback)
```

Comment:
Perhaps a better name is `onUpdate()` or `addCallback()`?

Reply:
Renamed to `onUpdate`.

I made the global & local position classes inherit from `Subscription` so they'll have this and other functions.